aten/src/ATen/cuda/CUDAGeneratorImpl.cpp

```diff
@@ -93,57 +106,106 @@ void CUDAGeneratorImpl::set_current_seed(uint64_t seed) {
  * seeds the CPUGeneratorImpl with it and then returns that number.
- * 
+ *
  * FIXME: You can move this function to Generator.cpp if the algorithm
  * in getNonDeterministicRandom is unified for both CPU and CUDA
  */
 uint64_t CUDAGeneratorImpl::seed() {
+  at::cuda::assertNotCapturing("Cannot call CUDAGeneratorImpl::seed");
   auto random = c10::detail::getNonDeterministicRandom(true);
   this->set_current_seed(random);
   return random;
 }
 
 /**
  * Sets the philox_offset_per_thread_ to be used by curandStatePhilox4_32_10
- * 
+ *
  * See Note [Acquire lock when using random generators]
  */
 void CUDAGeneratorImpl::set_philox_offset_per_thread(uint64_t offset) {
+  at::cuda::assertNotCapturing("Cannot call CUDAGeneratorImpl::set_philox_offset_per_thread");
   philox_offset_per_thread_ = offset;
 }
 
 /**
  * Gets the current philox_offset_per_thread_ of CUDAGeneratorImpl.
  */
 uint64_t CUDAGeneratorImpl::philox_offset_per_thread() {
+  at::cuda::assertNotCapturing("Cannot call CUDAGeneratorImpl::philox_offset_per_thread");
   return philox_offset_per_thread_;
 }
 
+/**
+ * Prepares this instance for a cuda graph capture region.
+ * offset_extragraph is the initial offset at the start of the graphed region.
+ * offset_intragraph tracks the offset in the graphed region.
+ */
+void CUDAGeneratorImpl::graph_prologue(int64_t* offset_extragraph) {
+  TORCH_INTERNAL_ASSERT(is_default_);
+  offset_extragraph_ = offset_extragraph;
+  offset_intragraph_ = 0;
+}
+
+/**
+ * Finalizes a cuda graph capture region for this instance.
+ */
+uint64_t CUDAGeneratorImpl::graph_epilogue() {
+  TORCH_INTERNAL_ASSERT(is_default_);
+  return offset_intragraph_;
+}
+
 /**
  * Gets the seed and philox offset value to be used in
- * curandStatePhilox4_32_10
- * 
+ * curandStatePhilox4_32_10, in an opaque PhiloxCudaState that's safe
+ * and can be used non-divergently in callers whether CUDA graph
+ * capture is underway or not.  See
+ * Note [CUDA Graph-safe RNG states]
+ *
  * Each kernel using philox has to sensibly increment offset
  * for future users of philox. So it gets the "old" value for
  * itself (before add), and tells subsequent users which offset
  * they should use, since only the kernel knows how many randoms
- * it intends to generate. 
- * 
+ * it intends to generate.
+ *
  * Increment should be at least the number of curand() random numbers used in
  * each thread. It is the user's responsibility to make sure that the increment
  * for philox is never smaller than the number of curand() calls. Increment
  * value > the number of curand() calls won't harm but anything less would mean
  * that you would be reusing random values from previous calls.
- * 
+ *
  * See Note [Acquire lock when using random generators]
  */
+PhiloxCudaState CUDAGeneratorImpl::philox_cuda_state(uint64_t increment) {
+  if (at::cuda::currentStreamCaptureStatus()) {
+    TORCH_INTERNAL_ASSERT(is_default_);
```

Comment:
This shouldn't be an internal assert, because it's pretty easy for a user to trigger it (just use a non-default generator inside a capture region). So instead it should be a CHECK, and an error message saying how to fix your code to not have this problem / report a bug asking for generator support.

Reply:
hopefully resolved